File: src/executor.js
'use strict';
const { spawn, execSync } = require('child_process');
const os   = require('os');
const { session } = require('./config');

const platform = process.platform;

// Track cwd ourselves because child `cd` doesn't affect parent
let _cwd = process.cwd();

function getCwd() { return _cwd; }

// ─── CHECK WINDOWS ADMIN ─────────────────────────────────────────────────────
function isWindowsAdmin() {
  try {
    execSync('net session', { stdio: 'ignore' });
    return true;
  } catch {
    return false;
  }
}

// ─── ELEVATE ─────────────────────────────────────────────────────────────────
// Windows: relaunch ShellMax in a new elevated cmd window
// Mac/Linux: set adminMode flag (sudo prefixed on each command)
function elevate() {
  if (platform === 'win32') {
    if (isWindowsAdmin()) {
      return false; // already elevated
    }
    // Relaunch elevated – user will see UAC prompt
    const node  = process.execPath;
    const entry = require.resolve('../bin/shellmax.js');
    const ps    = `Start-Process -FilePath "${node}" -ArgumentList "${entry}" -Verb RunAs`;
    spawn('powershell.exe', ['-Command', ps], { detached: true, stdio: 'ignore' }).unref();
    return true; // caller should exit
  } else {
    // Mac / Linux: just flip the flag; sudo will prompt on first command
    session.adminMode = true;
    return false;
  }
}

// ─── BUILD SHELL INVOCATION ───────────────────────────────────────────────────
function buildShellArgs(input) {
  const trimmed = input.trim();

  if (trimmed.startsWith('ps:')) {
    // PowerShell command
    const psCmd = trimmed.slice(3).trim();
    if (platform === 'win32') {
      return { shell: 'powershell.exe', flag: '-Command', cmd: psCmd };
    } else {
      // Try pwsh (PowerShell Core on Mac/Linux)
      return { shell: 'pwsh', flag: '-Command', cmd: psCmd };
    }
  }

  // Normal command – apply sudo on Mac/Linux if adminMode
  let cmd = trimmed;
  if (session.adminMode && platform !== 'win32' && !cmd.startsWith('sudo ')) {
    cmd = `sudo ${cmd}`;
  }

  if (platform === 'win32') {
    return { shell: 'cmd.exe', flag: '/c', cmd };
  } else {
    return { shell: '/bin/bash', flag: '-c', cmd };
  }
}

// ─── RUN ──────────────────────────────────────────────────────────────────────
// Returns a Promise that resolves with output when command exits.
function run(input, rl) {
  const trimmed = input.trim();
  if (!trimmed) return Promise.resolve('');

  // Handle `cd` ourselves
  if (/^cd(\s|$)/.test(trimmed)) {
    const parts  = trimmed.split(/\s+/);
    const target = parts[1] ? parts[1].replace(/^["']|["']$/g, '') : os.homedir();
    try {
      const resolved = target.replace('~', os.homedir());
      process.chdir(resolved);
      _cwd = process.cwd();
    } catch (err) {
      return Promise.resolve(`shellmax: cd: ${err.message}`);
    }
    return Promise.resolve('');
  }

  const { shell, flag, cmd } = buildShellArgs(trimmed);

  return new Promise((resolve) => {
    if (rl) rl.pause();

    let stdout = '';
    let stderr = '';

    const child = spawn(shell, [flag, cmd], {
      cwd:   _cwd,
      env:   process.env,
    });

    if (child.stdout) {
      child.stdout.on('data', (data) => {
        stdout += data.toString();
      });
    }

    if (child.stderr) {
      child.stderr.on('data', (data) => {
        stderr += data.toString();
      });
    }

    child.on('error', (err) => {
      if (rl) rl.resume();
      resolve(`shellmax: ${err.message}`);
    });

    child.on('close', () => {
      _cwd = process.cwd();
      if (rl) rl.resume();
      resolve(stdout + (stderr ? '\n' + stderr : ''));
    });
  });
}

module.exports = { run, elevate, isWindowsAdmin, getCwd };

File: src/executor.rs
use anyhow::Result;
use std::env;
use std::process::Command;
use std::sync::Mutex;

static CWD: Mutex<Option<String>> = Mutex::new(None);

fn get_cwd_inner() -> String {
    let mut cwd = CWD.lock().unwrap();
    cwd.clone().unwrap_or_else(|| {
        let current = env::current_dir()
            .ok()
            .map(|p| p.to_string_lossy().to_string())
            .unwrap_or_else(|| ".".into());
        *cwd = Some(current.clone());
        current
    })
}

pub fn get_cwd() -> String {
    get_cwd_inner()
}

pub fn run_shell(input: &str) -> Result<String> {
    let trimmed = input.trim().to_string();
    if trimmed.is_empty() {
        return Ok(String::new());
    }

    if trimmed.starts_with("cd ") || trimmed == "cd" {
        let target = trimmed[2..].trim();
        let target = if target.is_empty() {
            dirs::home_dir()
                .map(|p| p.to_string_lossy().to_string())
                .unwrap_or_else(|| ".".into())
        } else {
            target.replace(
                '~',
                &dirs::home_dir()
                    .map(|p| p.to_string_lossy().to_string())
                    .unwrap_or_default(),
            )
        };
        match env::set_current_dir(&target) {
            Ok(_) => {
                let mut cwd = CWD.lock().unwrap();
                *cwd = Some(env::current_dir().unwrap().to_string_lossy().to_string());
                Ok(String::new())
            }
            Err(e) => Ok(format!("shellmax: cd: {}", e)),
        }
    } else {
        let (shell, flag) = if cfg!(target_os = "windows") {
            ("cmd.exe", "/C")
        } else {
            ("/bin/bash", "-c")
        };
        let output = Command::new(shell)
            .arg(flag)
            .arg(&trimmed)
            .current_dir(get_cwd_inner())
            .output()?;
        let mut result = String::from_utf8_lossy(&output.stdout).to_string();
        if !output.stderr.is_empty() {
            if !result.is_empty() {
                result.push('\n');
            }
            result.push_str(&String::from_utf8_lossy(&output.stderr));
        }
        Ok(result.trim().to_string())
    }
}
